### DISCUSSION TITLE
Feature request: add a hook to do side effect logic when loading translation failed

### DISCUSSION THREAD

**Discussion by ahnpnl** [robisim74/qwik-speak#45] [Ideas]
I would like to have a hook to do side effect when failing to load translations. The purpose is I want to separate the concern between the `loadTranslation$`  and side effect logic I want to execute.

An example of the hook can be:
```
useOnFailedTranslations((error) => {
      loggingService.log('Write logging', error);
});
```

This hook will execute whenever an error is thrown in `loadTranslation$`

**Comment by robisim74** [comment]
It's not clear why you would run an effect outside the `loadTranslation$`: what does your function look like and what would be the use cases?

**Reply by ahnpnl** [reply]
My function is the same as doc
```
const loadTranslation$: LoadTranslationFn = server$((lang: string, asset: string) => {
  const langAsset = `/i18n/${lang}/${asset}.json`;
  if (langAsset in translationData) {
    return translationData[langAsset]();
  }
  if (isDev) {
    console.warn(`loadTranslation$: ${langAsset} not found`);
  }
  return null;
});
```

My team would like to have a clearer separating of concern that this function should only load the translations. If putting the logging into this function, it goes against the purpose of single responsibility. 

To make it a bit more extended, later the `console.warn` would be replace by a logging service which integrates with Cloud logging service for example.

**Comment by robisim74** [comment]
A `Cloud logging service` in dev mode?

In production, using the above function, a js chunk is created in `server/assets` for each translation file, and the `Inline `plugin uses those same files to inline the translations.

If any files are missing, your production build is wrong: do you want to find out with a logger once it's published?

**Reply by ahnpnl** [reply]
I think a logger would be still nice, I think it's more like an edge case of such case can happen.

**Reply by robisim74** [reply]
I don't like the idea of a hook just for one function, at this point I would like to think about a hook for all library errors.

Do you know any similar cases?

**Reply by ahnpnl** [reply]
I think a hook for all errors would be the best case. As far as I can see from `qwik-speak`, I see a few scenarios: 
- cannot find translation file 
- warning about translation keys don't exist
- warning about requested locale isn't supported

The first 2 cases can happen. The last one I'm not sure if we can log it as well. Not sure if the hook should call when there are such warnings?

**Reply by robisim74** [reply]
I was wondering if you know any other libraries that expose a similar hook, to have a reference on how to pass/map errors

**Reply by ahnpnl** [reply]
Hmm for Qwik no, `qwik-speak` is the only library I'm using right now 😄 

I think the closest example would be https://react-hook-form.com/api/useformstate/ . That hook returns the whole form state which contains also error object.

**Reply by robisim74** [reply]
Ok, I have to evaluate... maybe it could be called `useSpeakError` and meanwhile include the errors of `loadTranslation$`. By the way: if you use this hook, then the consumer has to give up implementing error handling inside the `loadTranslation$`

**Reply by ahnpnl** [reply]
Qwik is still at RC phase so I think there is still time to have such change for API :)

**Reply by robisim74** [reply]
It is not a problem to change the APIs. My doubt is this: if you handle the error inside` loadTranslation$`, you can for example pass a fallback translation. With the hook instead you have to give up handling the error inside the function and no fallback will be possible.

**Reply by ahnpnl** [reply]
Hmm that’s true. If we make `loadTranslation` to return an object with translations and error looks kinda ugly too. 

I think for fallback translation, should we control it via speakConfig rather than at loadTranslation?

**Reply by ahnpnl** [reply]
I see the loadTranslation has some arguments, maybe we can add new argument which is a callback? The callback will trigger an event hook?

**Comment by robisim74** [comment]
Uhm... _return an object with translations and error looks kinda ugly too_: When you make API servers, you always return an object and a code (ok, or error): it could be a good idea, because the hook would return the same code provided by the function and would be known by the consumer (also solving the problem of mappings).

Callback? never tried it in a Qwik QRL function...

**Reply by ahnpnl** [reply]
I’m fine with both options, curious about option 2 :)
